x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/AbstractRateAggregator.java

```diff
@@ -54,21 +57,42 @@ public AbstractRateAggregator(
         }
         this.rateUnit = rateUnit;
         this.rateMode = rateMode;
-        this.sizedBucketAggregator = findSizedBucketAncestor();
+        this.rateDivisorCalculator = new CachedSupplier<>(this::constructRateDivisor);
     }
 
-    private SizedBucketAggregator findSizedBucketAncestor() {
+    private RateDivisorCalculator constructRateDivisor() {
         SizedBucketAggregator sizedBucketAggregator = null;
+        boolean isParent = false;
         for (Aggregator ancestor = parent; ancestor != null; ancestor = ancestor.parent()) {
             if (ancestor instanceof SizedBucketAggregator) {
                 sizedBucketAggregator = (SizedBucketAggregator) ancestor;
+                isParent = ancestor == parent;
                 break;
             }
         }
         if (sizedBucketAggregator == null) {
-            throw new IllegalArgumentException("The rate aggregation can only be used inside a date histogram");
+            for (Aggregator ancestor = parent; ancestor != null; ancestor = ancestor.parent()) {
+                if (ancestor instanceof CompositeAggregator) {
+                    List<SizedBucketAggregator> compositeSizedBucketAggregators = ((CompositeAggregator) ancestor)
+                        .getSizedBucketAggregators();
+                    if (compositeSizedBucketAggregators.size() != 1) {
```

Comment:
I wonder if it would be cleaner to add a `.getSizedBucketAggregator()` method instead and hide all this logic inside this method. We pretty much only care if it can return size or not.

Reply:
@imotov the issue here is WHICH date_histogram if multiple are within the composite aggregation. 

If `composite` satisfies the `SizedBucketAggregator` interface, which `date_histogram` source should it pick (if it has one)?

It could simply be the first one in the sources array. Which seems OK to me.

Reply:
I am ok with the current logic, I am just proposing to hid this logic inside SizedBucketAggregator